google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/gcs/GcsBucketContentEditorPanel.java

```diff
@@ -17,36 +17,69 @@
 package com.google.cloud.tools.intellij.gcs;
 
 import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.Bucket;
+import com.google.cloud.storage.Storage.BlobListOption;
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.Iterators;
-import java.util.Arrays;
+import com.google.common.collect.Lists;
+import java.awt.event.MouseAdapter;
+import java.awt.event.MouseEvent;
 import java.util.List;
-import java.util.Vector;
-import java.util.stream.Collectors;
-import java.util.stream.StreamSupport;
 import javax.swing.JPanel;
 import javax.swing.JTable;
-import javax.swing.table.DefaultTableModel;
 import org.jetbrains.annotations.NotNull;
 
 /** Defines the Google Cloud Storage bucket content browsing UI panel. */
-// TODO(eshaul) work in progress
+// TODO(eshaul) implement toolbar panel and breadcrumb browser
 final class GcsBucketContentEditorPanel {
 
+  private Bucket bucket;
+
   private JPanel bucketContentEditorPanel;
   private JPanel bucketContentToolbarPanel;
   private JPanel breadCrumbsPanel;
   private JTable bucketContentTable;
+  private GcsBlobTableModel tableModel;
+
+  GcsBucketContentEditorPanel(@NotNull Bucket bucket) {
+    this.bucket = bucket;
 
-  private static final List<String> TABLE_COL_HEADER =
-      Arrays.asList("Name", "Size", "Type", "Last Modified");
+    bucketContentTable.addMouseListener(
+        new MouseAdapter() {
+          @Override
+          public void mouseClicked(MouseEvent event) {
+            if (event.getClickCount() == 2 && tableModel != null) {
+              Blob selectedBlob =
+                  tableModel.getBlobAt(bucketContentTable.rowAtPoint(event.getPoint()));
+
+              if (selectedBlob.isDirectory()) {
+                updateTableModel(selectedBlob.getName());
+              }
+            }
+          }
+        });
+  }
 
-  void setTableModel(@NotNull Iterable<Blob> blobs) {
-    if (Iterators.size(blobs.iterator()) != 0) {
-      bucketContentTable.setModel(new GcsBucketTableModel(blobs));
+  void initTableModel() {
+    List<Blob> blobs = getBlobsAtLevel("");
+    if (!blobs.isEmpty()) {
+      tableModel = new GcsBlobTableModel();
+      tableModel.setDataVector(blobs, "");
+      bucketContentTable.setModel(tableModel);
     }
   }
 
+  private void updateTableModel(String prefix) {
+    tableModel.setRowCount(0);
+    tableModel.setDataVector(getBlobsAtLevel(prefix), prefix);
+    tableModel.fireTableDataChanged();
+  }
+
+  // TODO(eshaul) should be done asynchronously and show loader in the UI
+  private List<Blob> getBlobsAtLevel(String prefix) {
```

Comment:
I'm not sure what "level" refers to here. Maybe this should be renamed to `getBlobsForPrefix()`?

Reply:
I updated the name to `getBlobsStartingWith` and added javadoc to explain whats going on. Let me know if its more clear now.